src/v0.3/concatenation.jl

```diff
@@ -0,0 +1,162 @@
+import Base: promote_rule, promote_type, cat_t, hcat, vcat, hvcat, cat
+
+type DummyJuMPArray end
+
+Base.promote_rule{T<:OneIndexedArray}(::Type{T},::Type{Union()}) = DummyJuMPArray
+Base.promote_rule{T<:OneIndexedArray,S<:OneIndexedArray}(::Type{T},::Type{S}) = DummyJuMPArray
+Base.promote_rule{T<:OneIndexedArray,S}(::Type{T},::Type{S}) = DummyJuMPArray
+
+Base.promote_type{T<:OneIndexedArray}(::Type{T},::Type{Union()}) = DummyJuMPArray
+Base.promote_type{T<:OneIndexedArray}(::Type{Union()},::Type{T}) = DummyJuMPArray
+Base.promote_type{T<:OneIndexedArray}(::Type{T},::Type{T}) = DummyJuMPArray
+Base.promote_type{T<:OneIndexedArray,S<:OneIndexedArray}(::Type{T},::Type{S}) = DummyJuMPArray
+Base.promote_type{T<:OneIndexedArray,S}(::Type{T},::Type{S}) = DummyJuMPArray
+
+AffExpr(v::Variable) = AffExpr(Variable[v], Float64[1], 0.0)
+AffExpr(aff::AffExpr) = copy(aff)
+QuadExpr(v::Variable) = QuadExpr(Variable[], Variable[], Float64[], AffExpr(v))
+QuadExpr(aff::AffExpr) = QuadExpr(Variable[], Variable[], Float64[], copy(aff))
+QuadExpr(q::QuadExpr) = copy(q)
+
+_tofull(x) = x
+_tofull(x::OneIndexedArray) = x.innerArray
+
+Base.hcat(X::OneIndexedArray...) = hcat([_tofull(x) for x in X]...)
+Base.vcat(X::OneIndexedArray...) = vcat([_tofull(x) for x in X]...)
+Base.hvcat(rows::(Int...), X::OneIndexedArray...) = hvcat(rows, [_tofull(x) for x in X]...)
+
+typealias BaseTypes Union(Number,AbstractArray)
+typealias CatTypes Union(Number,JuMPTypes,AbstractArray)
+typealias CatTypesFull Union(CatTypes,OneIndexedArray)
+
+Base.vcat(X::CatTypes...) = Base.cat(1, X...)
+Base.hcat(X::CatTypes...) = Base.cat(2, X...)
+
+###############################################################################
+# copied from Base at https://github.com/JuliaLang/julia/blob/0c24dca65c031820b91721139f0291068086955c/base/abstractarray.jl#L613-L668
```

Comment:
We need to include a copyright and license notice if we're copying code from julia. It can just be in a comment right here.